File: backend/flowforge-common/src/main/java/com/flowforge/common/validator/DagValidator.java
package com.flowforge.common.validator;

import com.flowforge.common.model.NodeCategory;
import com.flowforge.common.model.WorkflowEdge;
import com.flowforge.common.model.WorkflowGraph;
import com.flowforge.common.model.WorkflowNode;

import java.util.*;

public class DagValidator {

    public static ValidationResult validate(WorkflowGraph graph) {
        List<String> errors = new ArrayList<>();
        List<String> warnings = new ArrayList<>();

        if (graph == null || graph.getNodes() == null || graph.getNodes().isEmpty()) {
            return ValidationResult.failure(List.of("Workflow must contain at least one node."));
        }

        List<WorkflowNode> nodes = graph.getNodes();
        List<WorkflowEdge> edges = graph.getEdges() != null ? graph.getEdges() : Collections.emptyList();

        Map<String, WorkflowNode> nodeMap = new HashMap<>();
        Set<String> triggerNodeIds = new HashSet<>();

        // 1. Verify Node ID uniqueness and identify triggers
        for (WorkflowNode node : nodes) {
            if (node.getId() == null || node.getId().trim().isEmpty()) {
                errors.add("Every node must have a non-empty ID.");
                continue;
            }
            if (nodeMap.containsKey(node.getId())) {
                errors.add("Duplicate node ID detected: " + node.getId());
            }
            if (node.getType() == null) {
                errors.add("Node '" + node.getId() + "' is missing a node type.");
            } else if (node.getType().getCategory() == NodeCategory.TRIGGER) {
                triggerNodeIds.add(node.getId());
            }
            nodeMap.put(node.getId(), node);
        }

        if (triggerNodeIds.isEmpty()) {
            errors.add("Workflow must have at least one trigger node (e.g. TRIGGER_MANUAL, TRIGGER_WEBHOOK, TRIGGER_SCHEDULE).");
        }

        // 2. Validate Edges
        Map<String, List<String>> adjacencyList = new HashMap<>();
        Map<String, Integer> inDegree = new HashMap<>();
        for (String nodeId : nodeMap.keySet()) {
            adjacencyList.put(nodeId, new ArrayList<>());
            inDegree.put(nodeId, 0);
        }

        for (WorkflowEdge edge : edges) {
            if (edge.getSourceNodeId() == null || edge.getTargetNodeId() == null) {
                errors.add("Edge must specify both sourceNodeId and targetNodeId.");
                continue;
            }
            if (!nodeMap.containsKey(edge.getSourceNodeId())) {
                errors.add("Edge references non-existent source node: " + edge.getSourceNodeId());
                continue;
            }
            if (!nodeMap.containsKey(edge.getTargetNodeId())) {
                errors.add("Edge references non-existent target node: " + edge.getTargetNodeId());
                continue;
            }
            if (edge.getSourceNodeId().equals(edge.getTargetNodeId())) {
                errors.add("Self-loop edge detected on node: " + edge.getSourceNodeId());
                continue;
            }

            // Triggers should not have incoming edges
            if (triggerNodeIds.contains(edge.getTargetNodeId())) {
                errors.add("Trigger node '" + edge.getTargetNodeId() + "' cannot have incoming edges.");
            }

            adjacencyList.get(edge.getSourceNodeId()).add(edge.getTargetNodeId());
            inDegree.put(edge.getTargetNodeId(), inDegree.get(edge.getTargetNodeId()) + 1);
        }

        // 3. Cycle Detection using Kahn's Algorithm
        Queue<String> queue = new LinkedList<>();
        for (Map.Entry<String, Integer> entry : inDegree.entrySet()) {
            if (entry.getValue() == 0) {
                queue.add(entry.getKey());
            }
        }

        int visitedCount = 0;
        while (!queue.isEmpty()) {
            String current = queue.poll();
            visitedCount++;

            for (String neighbor : adjacencyList.get(current)) {
                int updatedInDegree = inDegree.get(neighbor) - 1;
                inDegree.put(neighbor, updatedInDegree);
                if (updatedInDegree == 0) {
                    queue.add(neighbor);
                }
            }
        }

        if (visitedCount < nodeMap.size()) {
            errors.add("Cycle detected in workflow graph! Workflows must be Directed Acyclic Graphs (DAG).");
        }

        // 4. Check for Orphaned / Unreachable nodes
        for (WorkflowNode node : nodes) {
            if (!triggerNodeIds.contains(node.getId())) {
                // If it's not a trigger and has inDegree == 0 initially, check whether it was ever reached
                boolean hasIncoming = edges.stream().anyMatch(e -> node.getId().equals(e.getTargetNodeId()));
                if (!hasIncoming) {
                    warnings.add("Node '" + node.getId() + "' (" + node.getName() + ") has no incoming connections and is unreachable.");
                }
            }
        }

        boolean isValid = errors.isEmpty();
        return ValidationResult.builder()
                .valid(isValid)
                .errors(errors)
                .warnings(warnings)
                .build();
    }
}

File: frontend/src/utils/dagValidator.ts
import type { Node, Edge } from '@xyflow/react';
import type { ValidationResult, WorkflowNodeData } from '../types/workflow';

export function validateDag(nodes: Node<WorkflowNodeData>[], edges: Edge[]): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (nodes.length === 0) {
    return {
      valid: false,
      errors: ['The canvas is empty. Add at least one trigger node to start your workflow.'],
      warnings: [],
    };
  }

  const nodeMap = new Map<string, Node<WorkflowNodeData>>();
  const triggerIds = new Set<string>();

  for (const node of nodes) {
    nodeMap.set(node.id, node);
    if (node.data.category === 'TRIGGER') {
      triggerIds.add(node.id);
    }
  }

  if (triggerIds.size === 0) {
    errors.push('Workflow must contain at least one Trigger node (e.g. Webhook, Schedule, or Manual Trigger).');
  }

  // Build Adjacency List & inDegree
  const adjacencyList = new Map<string, string[]>();
  const inDegree = new Map<string, number>();

  for (const node of nodes) {
    adjacencyList.set(node.id, []);
    inDegree.set(node.id, 0);
  }

  for (const edge of edges) {
    if (!nodeMap.has(edge.source)) {
      errors.push(`Edge connects from an unknown node: ${edge.source}`);
      continue;
    }
    if (!nodeMap.has(edge.target)) {
      errors.push(`Edge connects to an unknown node: ${edge.target}`);
      continue;
    }

    if (edge.source === edge.target) {
      errors.push(`Self-loop detected on node "${nodeMap.get(edge.source)?.data.name}". Cycles are not allowed.`);
      continue;
    }

    if (triggerIds.has(edge.target)) {
      errors.push(`Trigger "${nodeMap.get(edge.target)?.data.name}" cannot have incoming connections.`);
    }

    adjacencyList.get(edge.source)!.push(edge.target);
    inDegree.set(edge.target, (inDegree.get(edge.target) || 0) + 1);
  }

  // Kahn's Algorithm for cycle detection
  const queue: string[] = [];
  for (const [nodeId, deg] of inDegree.entries()) {
    if (deg === 0) {
      queue.push(nodeId);
    }
  }

  let visitedCount = 0;
  while (queue.length > 0) {
    const current = queue.shift()!;
    visitedCount++;

    for (const neighbor of adjacencyList.get(current) || []) {
      const updated = inDegree.get(neighbor)! - 1;
      inDegree.set(neighbor, updated);
      if (updated === 0) {
        queue.push(neighbor);
      }
    }
  }

  if (visitedCount < nodes.length) {
    errors.push('Cycle detected in workflow! Directed Acyclic Graphs (DAG) cannot contain infinite loops.');
  }

  // Check for orphan non-trigger nodes
  for (const node of nodes) {
    if (!triggerIds.has(node.id)) {
      const hasIncoming = edges.some(e => e.target === node.id);
      if (!hasIncoming) {
        warnings.push(`Node "${node.data.name}" has no incoming connection and will not be triggered.`);
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}
